python/ray/serve/_private/autoscaling_state.py

```diff
@@ -301,6 +314,86 @@ def get_total_num_requests(self) -> float:
         return total_requests
 
 
+class ApplicationAutoscalingState:
+    """Manages autoscaling for a single application."""
+
+    def __init__(
+        self,
+        app_name: ApplicationName,
+        deployment_autoscaling_states: Dict[DeploymentID, DeploymentAutoscalingState],
+    ):
+        self._app_name = app_name
+        self._deployment_autoscaling_states = deployment_autoscaling_states
+        self._config: ServeApplicationSchema = None
+        self._policy: AutoscalingPolicy = None
+        self._policy_state: Dict[str, Dict] = None
+
+    def register(
+        self,
+        config: ServeApplicationSchema,
+    ):
+        self._config = config
+        self._policy = self._config.get_autoscaling_policy()
+        self._policy_state = {}
+
+    def get_scaling_decisions(
+        self, deployments: Dict[str, DeploymentDetails], _skip_bound_check: bool = False
+    ) -> Dict[str, int]:
+        autoscaling_contexts: Dict[str, AutoscalingContext] = {}
+        decisions: Dict[str, int] = {}
+        for name, deployment_detail in deployments.items():
+            deployment_id: DeploymentID = DeploymentID(
+                name=name, app_name=self._app_name
+            )
+            deployment_autoscaling_state = self._deployment_autoscaling_states[
+                deployment_id
+            ]
+            autoscaling_context: AutoscalingContext = AutoscalingContext(
+                deployment_id=deployment_id,
+                deployment_name=name,
+                app_name=self._app_name,
+                current_num_replicas=len(
+                    deployment_autoscaling_state._running_replicas
+                ),
+                target_num_replicas=deployment_detail.target_num_replicas,
+                running_replicas=deployment_autoscaling_state._running_replicas,
+                total_num_requests=deployment_autoscaling_state.get_total_num_requests(),
+                capacity_adjusted_min_replicas=deployment_autoscaling_state.get_num_replicas_lower_bound(),
+                capacity_adjusted_max_replicas=deployment_autoscaling_state.get_num_replicas_upper_bound(),
+                policy_state=self._policy_state.copy(),
+                current_time=time.time(),
+                config=deployment_autoscaling_state._config,
+                queued_requests=None,
+                requests_per_replica=None,
+                aggregated_metrics=None,
+                raw_metrics=None,
+                last_scale_up_time=None,
+                last_scale_down_time=None,
+            )
+
+            autoscaling_contexts[name] = autoscaling_context
+
+        decisions, self._policy_state = self._policy(autoscaling_contexts)
+
+        updated_decisions = {}
+
+        for deployment, decision_num_replicas in decisions.items():
+            deployment_id: DeploymentID = DeploymentID(
+                name=deployment, app_name=self._app_name
+            )
```

Comment:
let's make `decisions` `Dict[DeploymentID, int]`. User function will already have the DeploymentID in the context.

Reply:
updated.